File: assignment 7/c/programs/2.c
#include <stdio.h>
#include <stdlib.h>

struct node
{
    int info;
    struct node *ptr;
} *front, *rear, *temp;

void enq(int data);
void deq();
void display();

int main()
{
    int choice, data;
    front = rear = NULL;

    while (1)
    {
        printf("\n1. Enqueue\n2. Dequeue\n3. Display\n4. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice)
        {
        case 1:
            printf("Enter data to be enqueued: ");
            scanf("%d", &data);
            enq(data);
            break;
        case 2:
            deq();
            break;
        case 3:
            display();
            break;
        case 4:
            exit(0);
        default:
            printf("Invalid choice!\n");
        }
    }

    return 0;
}

void enq(int data)
{
    if (rear == NULL)
    {
        rear = (struct node *)malloc(sizeof(struct node));
        rear->ptr = NULL;
        rear->info = data;
        front = rear;
    }
    else
    {
        temp = (struct node *)malloc(sizeof(struct node));
        rear->ptr = temp;
        temp->info = data;
        temp->ptr = NULL;
        rear = temp;
    }
}

void deq()
{
    if (front == NULL)
    {
        printf("Queue is empty!\n");
    }
    else
    {
        temp = front;
        front = front->ptr;

        if (front == NULL)
        {
            rear = NULL;
        }

        printf("Dequeued element: %d\n", temp->info);
        free(temp);
    }
}

void display()
{
    if (front == NULL)
    {
        printf("Queue is empty!\n");
    }
    else
    {
        temp = front;

        while (temp != NULL)
        {
            printf("%d ", temp->info);
            temp = temp->ptr;
        }

        printf("\n");
    }
}

File: assignment 7/python/programs/2.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinearQueue:
    def __init__(self):
        self.front = self.rear = None

    def is_empty(self):
        return self.front is None

    def enqueue(self, item):
        new_node = Node(item)
        if self.rear is None:
            self.front = self.rear = new_node
            return
        self.rear.next = new_node
        self.rear = new_node

    def dequeue(self):
        if self.is_empty():
            print("Queue is empty. Cannot dequeue.")
            return
        item = self.front.data
        self.front = self.front.next
        if self.front is None:
            self.rear = None
        print(f"Dequeued: {item}")
        return item

    def display(self):
        if self.is_empty():
            print("Queue is empty.")
        else:
            current = self.front
            while current:
                print(current.data, end=" ")
                current = current.next
            print()


# Example usage:
if __name__ == "__main__":
    queue = LinearQueue()

    queue.enqueue(10)
    queue.enqueue(20)
    queue.enqueue(30)
    queue.enqueue(40)
    queue.enqueue(50)
    queue.display()

    queue.dequeue()
    queue.dequeue()
    queue.display()

    queue.enqueue(60)
    queue.enqueue(70)
    queue.display()
